### ISSUE TITLE
ATC 25 | Resource Multiplexing in Tuning and Serving Large Language Models

### ISSUE THREAD

**Issue by FortuneBush** [AI-Infra-Team/awesome-papers#76]
Title: Resource Multiplexing in Tuning and Serving Large Language Models
Authors: Yongjun He and Haofeng Yang, ETH Zurich; Yao Lu, National University of Singapore; Ana Klimovic and Gustavo Alonso, ETH Zurich
Paper: https://www.usenix.org/system/files/[atc25-he-yongjun.pdf](https://www.usenix.org/system/files/atc25-he-yongjun.pdf)

#### Motivation:
- Large Language Models (LLMs) often leave GPUs underutilized during tuning and inference due to varied resource demands in different phases (e.g., memory-bound decoding).
- Existing systems either avoid co-locating tuning and serving due to SLO concerns, or suffer from high context-switching overhead and interference.
- Parameter-Efficient Fine-Tuning (PEFT), like LoRA, opens opportunities for co-executing tuning and inference tasks by sharing base models and balancing compute/memory workloads.
<img width="888" height="259" alt="Image" src="https://github.com/user-attachments/assets/b789eaa4-eb70-404f-a1b2-0dc7bf6e1fab" />

#### Method:

<img width="300" height="208" alt="Image" src="https://github.com/user-attachments/assets/f9d3596c-3fd6-4382-a0e4-a1501716afb0" />

- **Iteration-level multitasking scheduler**: The scheduler dynamically determines how many fine-tuning “tasklets” can be safely co-executed with inference tasks in each iteration. It uses profiling data, cached runtime records, and a learned latency predictor to estimate task runtimes and prevent service-level objective (SLO) violations. 
- **Suspendable Autograd engine**: LLMStation transforms PyTorch’s backward process into coroutine-based segments called tasklets. These tasklets can be suspended and resumed flexibly, depending on current inference load and latency budgets.

<img width="400" height="282" alt="Image" src="https://github.com/user-attachments/assets/8195ca30-d383-43c6-9533-9db3f2cc0540" />

- **Fusion engine**: The Fusion engine combines memory-bound inference operations and compute-intensive fine-tuning forward passes within the same Transformer layer execution

<img width="350" height="250" alt="Image" src="https://github.com/user-attachments/assets/0f051876-c28c-4eb3-b6d1-9f2d0ee585c6" />

- **Shared memory manager**: The memory manager coordinates shared use of the base model weights, LoRA adapters, and intermediate states between fine-tuning and inference tasks. 

#### Experiments 

- LLMStation improves PEFT throughput by **1.38× to 14.77×** over state-of-the-art baselines (FineInfer, vLLM+Torchtune, chunked-training) while respecting SLOs.

<img width="1016" height="222" alt="Image" src="https://github.com/user-attachments/assets/0dd10e80-9afa-4ee6-8044-4402aae24050" />

- Achieves lower or comparable P99 TTFT and TPOT across real-world and synthetic workloads.

<img width="400" height="186" alt="Image" src="https://github.com/user-attachments/assets/97a348e7-5def-4867-92e2-b8248c51295b" />

- The scheduler adds negligible latency (≈18μs/iteration), and Autograd coroutines introduce <0.5% single-GPU overhead.
